python_modules/dagster-external/dagster_external/context.py

```diff
@@ -28,29 +30,41 @@
 def init_dagster_external() -> "ExternalExecutionContext":
     params = get_external_execution_params()
     if params.input_mode == ExternalExecutionIOMode.stdio:
-        input_stream = sys.stdin
+        with sys.stdin as f:
+            data = json.load(f)
     elif params.input_mode == ExternalExecutionIOMode.file:
         assert params.input_path, "input_path must be set when input_mode is `file`"
-        input_stream = open(params.input_path, "r")
+        with open(params.input_path, "r") as f:
+            data = json.load(f)
     elif params.input_mode == ExternalExecutionIOMode.fifo:
         assert params.input_path, "input_path must be set when input_mode is `fifo`"
-        input_stream = open(params.input_path, "r")
+        with open(params.input_path, "r") as f:
+            data = json.load(f)
+    elif params.input_mode == ExternalExecutionIOMode.socket:
+        assert params.host, "host must be set when input_mode is `socket`"
+        assert params.port, "port must be set when input_mode is `socket`"
+        with socket.create_connection((params.host, params.port)) as sock:
+            sock.makefile("w").write(f"{GET_CONTEXT_MESSAGE}\n")
+            data = json.loads(sock.makefile("r").readline())
+
```

Comment:
minor nit: could be cleaner if you extract this into a function

```python
def get_data():
    params = get_external_execution_params()
    if params.input_mode == ExternalExecutionIOMode.stdio:
        with sys.stdin as f:
            return json.load(f)
            
     # and so on and so forth
```

Even though Python allows it I find it weird to rely their loose scoping rules

Reply:
will address in the refactor